File: market/src/main/java/com/jingxi/smartlife/pad/market/mvp/view/AllServiceFragment.java
package com.jingxi.smartlife.pad.market.mvp.view;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.RecyclerView;

import com.jingxi.smartlife.pad.market.R;
import com.jingxi.smartlife.pad.market.mvp.adapter.FindServiceChildAdapter;
import com.jingxi.smartlife.pad.market.mvp.data.ServiceDto;
import com.jingxi.smartlife.pad.market.mvp.model.MarketModel;
import com.wujia.businesslib.data.ApiResponse;
import com.wujia.businesslib.data.CardDetailBean;
import com.wujia.businesslib.event.EventBusUtil;
import com.wujia.businesslib.event.EventSubscription;
import com.wujia.businesslib.event.IMiessageInvoke;
import com.wujia.lib.widget.HorizontalTabBar;
import com.wujia.lib_common.base.BasePresenter;
import com.wujia.lib_common.base.baseadapter.MultiItemTypeAdapter;
import com.wujia.lib_common.base.baseadapter.wrapper.LoadMoreWrapper;
import com.wujia.lib_common.base.view.ServiceCardDecoration;
import com.wujia.lib_common.data.network.SimpleRequestSubscriber;
import com.wujia.lib_common.data.network.exception.ApiException;
import com.wujia.lib_common.utils.LogUtil;
import com.wujia.lib_common.utils.ScreenUtil;

import java.util.ArrayList;

/**
 * author ：shenbingkai@163.com
 * date ：2019-02-17
 * description ：
 */
public class AllServiceFragment extends ServiceBaseFragment implements HorizontalTabBar.OnTabSelectedListener, LoadMoreWrapper.OnLoadMoreListener, SwipeRefreshLayout.OnRefreshListener, MultiItemTypeAdapter.OnRVItemClickListener {

    private RecyclerView recyclerView;
    private SwipeRefreshLayout mSwipeRefreshLayout;
    private boolean isLoading;
    private int pageSize = 12;
    private int pageNo = 1;
//    private int totleSize = 0;
    private ArrayList<CardDetailBean.ServicesBean> datas;
    private LoadMoreWrapper mLoadMoreWrapper;


    public static final String TYPE_ALL = "4";
    public static final String TYPE_MY = "1";
    public static final String TYPE_GOV = "3";

    private static final String KEY_TYPE = "type";

    String type = TYPE_MY;

    public void setType(String type) {
        this.type = type;
        if(mLoadMoreWrapper!=null){
            datas.clear();
            mLoadMoreWrapper.notifyDataSetChanged();
            mLoadMoreWrapper.setLoadMoreView(0);
            pageNo = 1;
            getList(true);
        }
    }

    MarketModel marketModel;


    private EventSubscription event = new EventSubscription(new IMiessageInvoke<EventSubscription>() {
        @Override
        public void eventBus(EventSubscription event) {
            if (event.getType() == EventSubscription.TYPE_NOTIFY) {
                pageNo = 1;
                getList(false);
            } else if (isVisible && event.eventType != EventSubscription.PUSH_NOTIFY) {
                mLoadMoreWrapper.notifyDataSetChanged();
            } else {
                switch (event.getType()) {
                    case EventSubscription.TYPE_GOV:
                        break;
                    case EventSubscription.TYPE_FIND:
                        if (type.equals(TYPE_GOV)) {
                            return;
                        }
                        break;

                }
                pageNo = 1;
                getList(true);
            }
        }
    });

    public AllServiceFragment() {

    }

    public static AllServiceFragment newInstance(String type) {
        AllServiceFragment fragment = new AllServiceFragment();
//        Bundle args = new Bundle();
//        args.putString(KEY_TYPE, type);
//        fragment.setArguments(args);
        fragment.type=type;
        return fragment;
    }

    @Override
    protected int getLayoutId() {
        return R.layout.fragment_service_all;
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putString(KEY_TYPE,type);
    }

    @Override
    public void onLazyInitView(@Nullable Bundle savedInstanceState) {
        super.onLazyInitView(savedInstanceState);
        if(savedInstanceState!=null){
            type =savedInstanceState.getString(KEY_TYPE);
            LogUtil.i("Allservice from savedInstanceState type="+type);
        }

        mSwipeRefreshLayout = $(R.id.swipe_container);
        recyclerView = $(R.id.rv1);
        recyclerView.addItemDecoration(new ServiceCardDecoration(ScreenUtil.dip2px(84)));

        datas = new ArrayList<>();

        FindServiceChildAdapter mAdapter = getAdapter(datas);
        mLoadMoreWrapper = new LoadMoreWrapper(mAdapter);
        recyclerView.setAdapter(mLoadMoreWrapper);
        mLoadMoreWrapper.setOnLoadMoreListener(this);
        mSwipeRefreshLayout.setOnRefreshListener(this);
        mAdapter.setOnItemClickListener(this);
        marketModel = new MarketModel();
        getList(true);
//        if(type.equals(TYPE_MY)){
        EventBusUtil.register(event);
//        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
//        if(type.equals(TYPE_MY)){
        EventBusUtil.unregister(event);
//        }
    }

    boolean isVisible;

    @Override
    public void onSupportVisible() {
        super.onSupportVisible();
        isVisible = true;
    }

    @Override
    public void onSupportInvisible() {
        super.onSupportInvisible();
        isVisible = false;
    }

    private void getList(boolean isShowLoadingDialog) {
        isLoading = true;
        addSubscribe(marketModel.getServiceList(type, pageNo, pageSize).subscribeWith(new SimpleRequestSubscriber<ApiResponse<ServiceDto>>(this,
                new SimpleRequestSubscriber.ActionConfig(isShowLoadingDialog, SimpleRequestSubscriber.SHOWERRORMESSAGE)) {
            @Override
            public void onResponse(ApiResponse<ServiceDto> response) {
                super.onResponse(response);
                if (mSwipeRefreshLayout.isRefreshing())
                    mSwipeRefreshLayout.setRefreshing(false);
                isLoading = false;

                if (pageNo == 1)
                    datas.clear();

                datas.addAll(response.data.getPage().getContent());

                if (response.data.getPage().isLast()) {
                    mLoadMoreWrapper.setLoadMoreView(0);
                } else {
                    mLoadMoreWrapper.setLoadMoreView(R.layout.view_loadmore);
                }

                mLoadMoreWrapper.notifyDataSetChanged();
                pageNo++;
            }

            @Override
            public void onFailed(ApiException apiException) {
                super.onFailed(apiException);
                isLoading = false;
                if (mSwipeRefreshLayout.isRefreshing())
                    mSwipeRefreshLayout.setRefreshing(false);
            }
        }));
    }

    @Override
    public void onTabSelected(int position, int prePosition) {

    }


    @Override
    public void onLoadMoreRequested() {
        if (mSwipeRefreshLayout.isRefreshing() || isLoading)
            return;
        getList(false);
    }

    @Override
    public void onRefresh() {
        if (isLoading)
            return;
        mLoadMoreWrapper.setLoadMoreView(0);
        pageNo = 1;
        getList(false);
    }

    @Override
    public void onItemClick(@Nullable RecyclerView.Adapter adapter, RecyclerView.ViewHolder holder, int position) {
        toTarget(datas.get(position));
    }

    @Override
    protected BasePresenter createPresenter() {
        return null;
    }
}

File: market/src/main/java/com/jingxi/smartlife/pad/market/mvp/view/AllServiceFragment.kt
package com.jingxi.smartlife.pad.market.mvp.view

import android.os.Bundle
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.jingxi.smartlife.pad.market.R
import com.jingxi.smartlife.pad.market.mvp.data.ServiceDto
import com.jingxi.smartlife.pad.market.mvp.model.MarketModel
import com.wujia.businesslib.data.ApiResponse
import com.wujia.businesslib.data.CardDetailBean
import com.wujia.businesslib.event.EventBusUtil
import com.wujia.businesslib.event.EventSubscription
import com.wujia.businesslib.event.IMiessageInvoke
import com.wujia.lib.widget.HorizontalTabBar
import com.wujia.lib_common.base.BasePresenter
import com.wujia.lib_common.base.BaseView
import com.wujia.lib_common.base.baseadapter.MultiItemTypeAdapter
import com.wujia.lib_common.base.baseadapter.wrapper.LoadMoreWrapper
import com.wujia.lib_common.base.view.ServiceCardDecoration
import com.wujia.lib_common.data.network.SimpleRequestSubscriber
import com.wujia.lib_common.data.network.exception.ApiException
import com.wujia.lib_common.utils.LogUtil
import com.wujia.lib_common.utils.ScreenUtil
import java.util.*

/**
 * author ：shenbingkai@163.com
 * date ：2019-02-17
 * description ：
 */
open class AllServiceFragment : ServiceBaseFragment<BasePresenter<BaseView>>(), HorizontalTabBar.OnTabSelectedListener, LoadMoreWrapper.OnLoadMoreListener, SwipeRefreshLayout.OnRefreshListener, MultiItemTypeAdapter.OnRVItemClickListener {
    override val layoutId: Int
        get() = R.layout.fragment_service_all
    private var recyclerView: RecyclerView? = null
    private var mSwipeRefreshLayout: SwipeRefreshLayout? = null
    private var isLoading: Boolean = false
    private val pageSize = 12
    private var pageNo = 1
    //    private int totleSize = 0;
    private var datas: ArrayList<CardDetailBean.ServicesBean>? = null
    private var mLoadMoreWrapper: LoadMoreWrapper? = null

     protected open var type: String= TYPE_ALL

    private val event = EventSubscription(object : IMiessageInvoke<EventSubscription> {
        override fun eventBus(event: EventSubscription) {
            if (event.type == EventSubscription.TYPE_NOTIFY) {
                pageNo = 1
                getList(false)
            } else if (isSupportVisible && event.eventType != EventSubscription.PUSH_NOTIFY) {
                mLoadMoreWrapper!!.notifyDataSetChanged()
            } else {
                pageNo = 1
                getList(true)
            }
        }
    })

    internal var isVisible: Boolean = false




    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putString(KEY_TYPE, type)
    }

    override fun onLazyInitView(savedInstanceState: Bundle?) {
        super.onLazyInitView(savedInstanceState)
        if (savedInstanceState != null) {
            type = savedInstanceState.getString(KEY_TYPE)!!
            LogUtil.i("Allservice from savedInstanceState type=" + type)
        }

        mSwipeRefreshLayout = `$`(R.id.swipe_container)
        recyclerView = `$`(R.id.rv1)
        recyclerView!!.addItemDecoration(ServiceCardDecoration(ScreenUtil.dip2px(84f)))

        datas = ArrayList()

        val mAdapter = getAdapter(datas!!)
        mLoadMoreWrapper = LoadMoreWrapper(mAdapter as RecyclerView.Adapter<RecyclerView.ViewHolder>)
        recyclerView!!.adapter = mLoadMoreWrapper
        mLoadMoreWrapper!!.setOnLoadMoreListener(this)
        mSwipeRefreshLayout!!.setOnRefreshListener(this)
        mAdapter.setOnItemClickListener(this)
        getList(true)
        //        if(type.equals(TYPE_MY)){
        EventBusUtil.register(event)
        //        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        //        if(type.equals(TYPE_MY)){
        EventBusUtil.unregister(event)
        //        }
    }

    override fun onSupportVisible() {
        super.onSupportVisible()
        isVisible = true
    }

    override fun onSupportInvisible() {
        super.onSupportInvisible()
        isVisible = false
    }

    private fun getList(isShowLoadingDialog: Boolean) {
        isLoading = true
        addSubscribe(MarketModel().getServiceList(type, pageNo, pageSize).subscribeWith(object : SimpleRequestSubscriber<ApiResponse<ServiceDto>>(this@AllServiceFragment,
                ActionConfig(isShowLoadingDialog, SHOWERRORMESSAGE)) {
            override fun onResponse(response: ApiResponse<ServiceDto>) {
                super.onResponse(response)
                if (mSwipeRefreshLayout!!.isRefreshing)
                    mSwipeRefreshLayout!!.isRefreshing = false
                isLoading = false

                if (pageNo == 1)
                    datas!!.clear()

                datas!!.addAll(response.data?.page!!.content!!)

                if (response.data!!.page!!.last) {
                    mLoadMoreWrapper!!.setLoadMoreView(0)
                } else {
                    mLoadMoreWrapper!!.setLoadMoreView(R.layout.view_loadmore)
                }

                mLoadMoreWrapper!!.notifyDataSetChanged()
                pageNo++
            }

            override fun onFailed(apiException: ApiException) {
                super.onFailed(apiException)
                isLoading = false
                if (mSwipeRefreshLayout!!.isRefreshing)
                    mSwipeRefreshLayout!!.isRefreshing = false
            }
        }))
    }

    override fun onTabSelected(position: Int, prePosition: Int) {

    }


    override fun onLoadMoreRequested() {
        if (mSwipeRefreshLayout!!.isRefreshing || isLoading)
            return
        getList(false)
    }

    override fun onRefresh() {
        if (isLoading)
            return
        mLoadMoreWrapper!!.setLoadMoreView(0)
        pageNo = 1
        getList(false)
    }

    override fun onItemClick(adapter: RecyclerView.Adapter<*>?, holder: RecyclerView.ViewHolder, position: Int) {
        toTarget(datas!![position])
    }

    override fun createPresenter(): BasePresenter<BaseView>? {
        return null
    }

    companion object {


       const val TYPE_ALL = "4"
       const val TYPE_MY = "1"
       const val TYPE_GOV = "3"

        private const val KEY_TYPE = "type"

    }

}
